executor/insert_common.go

```diff
@@ -368,102 +352,116 @@ func (e *InsertValues) getColDefaultValue(idx int, col *table.Column) (d types.D
 	return defaultVal, nil
 }
 
-// initDefaultValues fills generated columns, auto_increment column and empty column.
+// fillColValue fills the column value if it is not set in the insert statement.
+func (e *InsertValues) fillColValue(datum types.Datum, idx int, column *table.Column, hasValue bool) (types.Datum,
+	error) {
+	if mysql.HasAutoIncrementFlag(column.Flag) {
+		d, err := e.adjustAutoIncrementDatum(datum, hasValue, column)
+		if err != nil {
+			return types.Datum{}, errors.Trace(err)
+		}
+		return d, nil
+	}
+	if !hasValue {
+		d, err := e.getColDefaultValue(idx, column)
```

Comment:
if col is `generate` + `!hasValue`, this call can be eliminated?

Reply:
generated always doesn't have a value getColDefaultValue will return null for it.